guard/src/utils/reader.rs

```diff
@@ -0,0 +1,28 @@
+use std::fs::File;
+use std::io::{Cursor, Read, Stdin};
+
+pub struct Reader {
+    inner: ReadBuffer,
+}
+
+impl Read for Reader {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        match &mut self.inner {
+            ReadBuffer::Stdin(stdin) => stdin.read(buf),
+            ReadBuffer::Cursor(cursor) => cursor.read(buf),
+            ReadBuffer::File(file) => file.read(buf),
+        }
+    }
+}
+
+impl Reader {
+    pub fn new(stdin: ReadBuffer) -> Self {
+        Self { inner: stdin }
+    }
+}
+
+pub enum ReadBuffer {
+    Stdin(Stdin),
+    Cursor(Cursor<Vec<u8>>),
+    File(File),
```

Comment:
One enhancement idea (not for this PR) that we can take up later. For tests we are not planning to use reader at all we can add a NoOps reader that does nothing. We can do this for writer as well.